src/cdk/layout/breakpoints-observer.ts

```diff
@@ -0,0 +1,100 @@
+import {MediaMatcher} from './media-matcher';
+import {Observable} from 'rxjs/Observable';
+import {Subject} from 'rxjs/Subject';
+import {RxChain, map, startWith, takeUntil} from '@angular/cdk/rxjs';
+import {coerceArray} from '@angular/cdk/coercion';
+import {combineLatest} from 'rxjs/observable/combineLatest';
+import {fromEventPattern} from 'rxjs/observable/fromEventPattern';
+
+/** The current state of a layout breakpoint. */
+export interface BreakpointState {
+  query: string;
+  matches: boolean;
+}
+
+interface Query {
+  observable: Observable<BreakpointState>;
+  mql: MediaQueryList;
+}
+
+@Injectable()
+export class BreakpointObserver implements OnDestroy {
+  /**  A map of all media queries currently being listened for. */
+  private _queries: Map<string, Query> = new Map();
+  /** A subject for all other observables to takeUntil based on. */
+  private _activeSubject: Subject<{}> = new Subject();
+
+  constructor(private mediaMatcher: MediaMatcher, private zone: NgZone) {}
+
+  /** Completes the active subject, signalling to all other observables to complete. */
+  ngOnDestroy() {
+    this._activeSubject.next();
+    this._activeSubject.complete();
+  }
+
+  /** Whether the query currently is matched. */
+  isMatched(value: string | string[]): boolean {
+    let queries = coerceArray(value);
+    return queries.some((mediaQuery: string) => {
+      return this._registerQuery(mediaQuery).mql.matches;
+    });
+  }
+
+  /**
+   * Gets an observable of results for the given queries that will emit new results for any changes
+   * in matching of the given queries.
+   */
+  observe(value: string | string[]): Observable<BreakpointState> {
+    let queries = coerceArray(value);
+    let observables = queries.map(query => this._registerQuery(query).observable);
+
+    return combineLatest(observables, (a: BreakpointState, b: BreakpointState) => {
+      return <BreakpointState>{
+        matches: !!((a && a.matches) || (b && b.matches)),
+        query: queries.join(', ')
+      };
+    });
+  }
+
+  /** Registers a specific query to be listened for. */
+  private _registerQuery(query: string): Query {
+    // Only set up a new MediaQueryList if it is not already being listened for.
+    if (this._queries.has(query)) {
+      return this._queries.get(query)!;
+    }
+
+    let mql: MediaQueryList = this.mediaMatcher.matchMedia(query);
+    // Create callback for match changes and add it is as a listener.
+    let queryObservable = RxChain.from(fromEventPattern(
+      /**
+       * Listener callback methods are wrapped to be placed back in ngZone. Callbacks must be placed
+       * back into the zone because matchMedia is only included in Zone.js by loading the
+       * webapis-media-query.js file alongside the zone.js file.  Additionally, some browsers do not
+       * have MediaQueryList inherit from EventTarget, which causes inconsistencies in how Zone.js
+       * patches it.
+       */
```

Comment:
Use `//`-style comments for explanations / details (with `/** */` for class/member/variable descriptions)